File: educational-core/src/com/jetbrains/edu/coursecreator/actions/CCCreateCourseArchive.java
package com.jetbrains.edu.coursecreator.actions;

import com.intellij.ide.util.PropertiesComponent;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.actionSystem.CommonDataKeys;
import com.intellij.openapi.actionSystem.Presentation;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.project.DumbAwareAction;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.ui.DialogWrapper;
import com.intellij.openapi.vfs.VirtualFile;
import com.jetbrains.edu.coursecreator.CCUtils;
import com.jetbrains.edu.coursecreator.ui.CCCreateCourseArchiveDialog;
import com.jetbrains.edu.learning.StudyTaskManager;
import com.jetbrains.edu.learning.courseFormat.Course;
import com.jetbrains.edu.learning.courseFormat.EduCourse;
import com.jetbrains.edu.learning.courseFormat.ext.CourseExt;
import com.jetbrains.edu.learning.statistics.EduCounterUsageCollector;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;

import static com.intellij.openapi.ui.Messages.showErrorDialog;
import static com.jetbrains.edu.coursecreator.CCUtils.askToWrapTopLevelLessons;
import static com.jetbrains.edu.learning.EduUtils.addMnemonic;

@SuppressWarnings("ComponentNotRegistered") // educational-core.xml
public class CCCreateCourseArchive extends DumbAwareAction {

  public static final String GENERATE_COURSE_ARCHIVE = "Generate Course Archive";
  public static final String LAST_ARCHIVE_LOCATION = "Edu.CourseCreator.LastArchiveLocation";
  public static final String AUTHOR_NAME = "Edu.Author.Name";

  private String myZipName;
  private String myLocationDir;
  private String myAuthorName;

  public void setZipName(String zipName) {
    myZipName = zipName;
  }

  public void setLocationDir(String locationDir) {
    myLocationDir = locationDir;
  }

  public void setAuthorName(String authorName) {
    myAuthorName = authorName;
  }

  public CCCreateCourseArchive() {
    super(addMnemonic(GENERATE_COURSE_ARCHIVE), GENERATE_COURSE_ARCHIVE, null);
  }

  @Override
  public void update(@NotNull AnActionEvent e) {
    Presentation presentation = e.getPresentation();
    Project project = e.getProject();
    presentation.setEnabledAndVisible(project != null && CCUtils.isCourseCreator(project));
  }

  @Override
  public void actionPerformed(@NotNull AnActionEvent e) {
    final Project project = e.getData(CommonDataKeys.PROJECT);
    if (project == null) return;
    Course course = StudyTaskManager.getInstance(project).getCourse();
    assert course != null;

    if (CourseExt.getHasSections(course) && CourseExt.getHasTopLevelLessons(course)) {
      if (!askToWrapTopLevelLessons(project, (EduCourse)course)) {
        return;
      }
    }

    CCCreateCourseArchiveDialog dlg = new CCCreateCourseArchiveDialog(project, this);
    dlg.show();
    if (dlg.getExitCode() != DialogWrapper.OK_EXIT_CODE) {
      return;
    }
    course.setAuthorsAsString(new String[]{myAuthorName});
    String errorMessage = createCourseArchive(project, myZipName, myLocationDir, true);
    if (errorMessage == null) {
      PropertiesComponent.getInstance(project).setValue(LAST_ARCHIVE_LOCATION, myLocationDir);
      PropertiesComponent.getInstance(project).setValue(AUTHOR_NAME, myAuthorName);
      EduCounterUsageCollector.createCourseArchive();
    }
    else {
      showErrorDialog(project, errorMessage, "Failed to Create Course Archive");
    }
  }

  /**
   * @return null if course archive was created successfully, non-empty error message otherwise
   */
  @Nullable
  public static String createCourseArchive(@NotNull Project project, String zipName, String locationDir, boolean showMessage) {
    String archiveName = zipName + ".zip";
    VirtualFile jsonFolder = CCUtils.generateFolder(project, zipName);
    if (jsonFolder == null) {
      return "Failed to generate " + archiveName;
    }
    FileDocumentManager.getInstance().saveAllDocuments();
    final File zipFile = new File(locationDir, archiveName);
    return ApplicationManager.getApplication()
      .runWriteAction(new CourseArchiveCreator(project, jsonFolder, zipFile, showMessage));
  }
}

File: educational-core/src/com/jetbrains/edu/coursecreator/actions/CCCreateCourseArchive.kt
package com.jetbrains.edu.coursecreator.actions

import com.intellij.ide.actions.RevealFileAction
import com.intellij.ide.util.PropertiesComponent
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.actionSystem.CommonDataKeys
import com.intellij.openapi.application.*
import com.intellij.openapi.application.impl.coroutineDispatchingContext
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.openapi.progress.ProgressManager
import com.intellij.openapi.project.DumbAwareAction
import com.intellij.openapi.project.Project
import com.intellij.openapi.ui.Messages
import com.intellij.util.TimeoutUtil
import com.jetbrains.edu.coursecreator.CCNotificationUtils.showNotification
import com.jetbrains.edu.coursecreator.CCUtils.askToWrapTopLevelLessons
import com.jetbrains.edu.coursecreator.CCUtils.isCourseCreator
import com.jetbrains.edu.coursecreator.ui.CCCreateCourseArchiveDialog
import com.jetbrains.edu.learning.StudyTaskManager
import com.jetbrains.edu.learning.courseFormat.Course
import com.jetbrains.edu.learning.courseFormat.EduCourse
import com.jetbrains.edu.learning.courseFormat.Vendor
import com.jetbrains.edu.learning.courseFormat.ext.hasSections
import com.jetbrains.edu.learning.courseFormat.ext.hasTopLevelLessons
import com.jetbrains.edu.learning.getInEdt
import com.jetbrains.edu.learning.messages.EduCoreBundle
import com.jetbrains.edu.learning.statistics.EduCounterUsageCollector
import com.jetbrains.edu.learning.threadLocal
import io.ktor.utils.io.CancellationException
import kotlinx.coroutines.*
import org.jetbrains.annotations.NonNls
import java.io.File
import kotlin.concurrent.thread
import kotlin.coroutines.CoroutineContext

@Suppress("ComponentNotRegistered") // educational-core.xml
class CCCreateCourseArchive : AnAction(EduCoreBundle.lazyMessage("action.create.course.archive.text")) {

  override fun update(e: AnActionEvent) {
    val presentation = e.presentation
    val project = e.project
    presentation.isEnabledAndVisible = project != null && isCourseCreator(project)
  }

  override fun actionPerformed(e: AnActionEvent) {
    val project = e.getData(CommonDataKeys.PROJECT) ?: return
    val course = StudyTaskManager.getInstance(project).course ?: return

    if (course.hasSections && course.hasTopLevelLessons) {
      if (!askToWrapTopLevelLessons(project, (course as EduCourse))) {
        return
      }
    }

    CoroutineScope(Dispatchers.IO).launch {
      val flagCheckAllTasks = checkAllTasks(project, course)
      if (!flagCheckAllTasks) return@launch

      getInEdt {
        createSourceArchive(project, course)
      }
    }
  }

  /**
   * @return true if all tasks were checked successfully, false otherwise
   */
  private fun checkAllTasks(project: Project, course: Course): Boolean {
    val task = CheckAllTasksProgressTask(project, course)
    ProgressManager.getInstance().run(task)

    while (!task.isCancelled && !task.isFinished) { // waiting for background task to complete
      TimeoutUtil.sleep(200)
    }

    return when {
      task.isCancelled -> false
      (task.errorMessage == null) -> true
      else -> {
        getInEdt {
          Messages.showErrorDialog(project, task.errorMessage, EduCoreBundle.message("error.failed.to.create.course.archive"))
        }
        false
      }
    }
  }

  private fun createSourceArchive(project: Project, course: Course) {
    val dlg = CCCreateCourseArchiveDialog(project, course.name)
    if (!dlg.showAndGet()) {
      return
    }

    val locationPath = dlg.locationPath

    val authorName = dlg.authorName
    course.vendor = Vendor(authorName)
    PropertiesComponent.getInstance(project).setValue(AUTHOR_NAME, authorName)

    val errorMessage = createCourseArchive(project, locationPath)
    if (errorMessage == null) {
      showNotification(project, EduCoreBundle.message("action.create.course.archive.success.message"), ShowFileAction(locationPath))
      PropertiesComponent.getInstance(project).setValue(LAST_ARCHIVE_LOCATION, locationPath)
      EduCounterUsageCollector.createCourseArchive()
    } else {
      Messages.showErrorDialog(project, errorMessage, EduCoreBundle.message("error.failed.to.create.course.archive"))
    }
  }


  /**
   * @return null when course archive was created successfully, non-empty error message otherwise
   */
  private fun createCourseArchive(project: Project, location: String): String? {
    FileDocumentManager.getInstance().saveAllDocuments()
    return ApplicationManager.getApplication().runWriteAction<String>(CourseArchiveCreator(project, location))
  }

  companion object {
    @NonNls
    const val LAST_ARCHIVE_LOCATION = "Edu.CourseCreator.LastArchiveLocation"

    @NonNls
    const val AUTHOR_NAME = "Edu.Author.Name"
  }

  class ShowFileAction(val path: String) : AnAction(
    EduCoreBundle.message("action.create.course.archive.open.file", RevealFileAction.getFileManagerName())) {
    override fun actionPerformed(e: AnActionEvent) {
      RevealFileAction.openFile(File(path))
    }

    override fun update(e: AnActionEvent) {
      val presentation = e.presentation
      presentation.isVisible = RevealFileAction.isSupported()
    }
  }
}
